tests: bsim: Enable asserts by default

jori-nordic opened the pull request:
Asserts should always be enabled in test builds.

jori-nordic committed:
tests: bsim: Enable asserts by default

Asserts should always be enabled in test builds.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

aescolar commented:
Fair enough. Note the issues it found already.
CC @cvinayak @Thalley

Thalley commented:
> Fair enough. Note the issues it found already. CC @cvinayak @Thalley

Hmm, seems like it didn't even manage to start the audio tests :O

jori-nordic committed:
Bluetooth: tests: fix double-registering of callbacks

This was the source of assert failures.
If only they were enabled in the first place...

The issue is that in order to save on build time, these two tests build
only one image, from which two devices are
instantiated (`gatt_client_test.c` and `gatt_server_test.c`).

The double-registration happens as `BT_CONN_CB_DEFINE()` is build-time, and
is included in two files of the same image. Then _both_ images will end up
with both `connected()` `disconnected()` etc callbacks, which have logic
specific to their respective images/devices.

The patch uses runtime registration instead so each device only ever
registers its own callbacks and not the other device's too.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

cvinayak commented on tests/bsim/bluetooth/host/gatt/authorization/src/gatt_client_test.c:
```diff
@@ -297,6 +297,8 @@ static void test_main(void)
 {
 	int err;
 
+	bt_conn_cb_register(&conn_callbacks);
+
```
Explicitly register is required because? doesn't `BT_CONN_CB_DEFINE`  work? All samples use it, is there more to be fixed?

jhedberg replied:
> Explicitly register is required because? doesn't `BT_CONN_CB_DEFINE` work? All samples use it, is there more to be fixed?

The commit message mentions double registration, so I assume there was some place where the static registration caused problems. Admittedly the the commit message didn't really explain how/where this was happening, so I'm also curious to hear the answer to this.

alwa-nordic replied:
The commit message is misleading. The problem is not that anything got registered twice. It's that `gatt_client_test.c` and `gatt_server_test.c` are compiled in to the same image.

`gatt_client_test.c` is causing a problem:

`gatt_client_test.c:disconencted` does `bt_conn_unref(g_conn)`. `g_conn` is `NULL` in this situation, causing undefined behavior.

Note that `g_conn` is not set in `gatt_client_test.c:connected`.

When asserts are on, `bt_conn_unref(NULL)` panics.

This patch fixes the problem by only registering the problematic `gatt_client_test.c:disconencted` when running the test it was made for.

jori-nordic replied:
Yes I'm sorry. I have to stop writing like the reader is me :)
Commit message updated, thanks for translating my cryptic message @alwa-nordic .

jori-nordic replied:
It seems like CI will retrigger all builds and tests when only the commit message is updated. Noice.
